packages/owl-bot/src/database.ts

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import admin from 'firebase-admin';
+import {OwlBotLock} from './config-files';
+import {Configs, ConfigsStore} from './configs-store';
+
+export type Db = admin.firestore.Firestore;
+interface UpdateLockPr {
+  pullRequestId: string;
+}
+
+/**
+ * When firebase sees a / in a doc id, it interprets it as a collection name.
+ * So, we have to escape them.  Also escape +s because we use them for combining
+ * strings into keys.  And use a format that can be decoded by decodeURIComponent().
+ */
+export function encodeId(s: string): string {
+  return s.replace('%', '%25').replace('/', '%2F').replace('+', '%2B');
+}
+
+export function decodeId(s: string): string {
+  return decodeURIComponent(s);
+}
+
+function makeUpdateLockKey(repo: string, lock: OwlBotLock): string {
+  return [repo, lock.docker.image, lock.docker.digest].map(encodeId).join('+');
+}
+
+export class FirestoreConfigsStore implements ConfigsStore {
+  private db: Db;
+  readonly yamls: string;
+  readonly lock_update_prs: string;
+
+  /**
+   * @param collectionsPrefix should only be overridden in tests.
+   */
+  constructor(db: Db, collectionsPrefix = 'owl-bot-') {
+    this.db = db;
+    this.yamls = collectionsPrefix + 'yamls';
+    this.lock_update_prs = collectionsPrefix + 'lock-update-prs';
+  }
+
+  async getConfigs(repo: string): Promise<Configs | undefined> {
+    const docRef = this.db.collection(this.yamls).doc(encodeId(repo));
+    const doc = await docRef.get();
+    // Should we verify the data?
+    return doc.data() as Configs;
```

Comment:
I'm surprised there's no generic interface on `data<Configs>()` that lets you do this instead of a cast.  I try to avoid the casts when at all possible.

Reply:
I found no way to avoid the cast.